src/MonoMod.UnitTest/CalliTest.cs

```diff
@@ -0,0 +1,133 @@
+﻿using Mono.Cecil.Cil;
+using MonoMod.Core.Platforms;
+using MonoMod.Utils;
+using System;
+using Xunit;
+using Xunit.Abstractions;
+
+
+namespace MonoMod.UnitTest
+{
+    public unsafe sealed class CalliTests : TestBase
+    {
+        public CalliTests(ITestOutputHelper helper) : base(helper)
+        {
+        }
+
+        [Fact]
+        public void CalliS()
+        {
+            if (PlatformDetection.Runtime is RuntimeKind.Mono)
+            {
+                LightCalli<random_struct>();
+            }
+            else
+            {
+                Calli<random_struct>();
+            }
+        }
+        [Fact]
+        public void CalliO()
+        {
+            if (PlatformDetection.Runtime is RuntimeKind.Mono)
+            {
+                LightCalli<random_class>();
+            }
+            else
+            {
+                Calli<random_class>();
+            }
+        }
+        delegate T Helper<T>(T o, ref T r, T[] a, T* p) where T : new();
+        public static unsafe T Ret<T>(T o, ref T r, T[] a, T* p) where T : new()
+        {
+            Assert.Equal(typeof(T), o.GetType());
+            Assert.Equal(typeof(T[]), a.GetType());
+            Assert.Equal(r, *p);
+            r = new T();
+            Assert.Equal(r, *p);
+            return o;
+        }
+        unsafe void Calli<T>() where T : new()
+        {
+            T obj = new();
+            Ret(obj, ref obj, [obj], (T*)Unsafe.AsPointer(ref obj));
```

Comment:
`Unsafe.AsPointer` shouldn't be needed here. You should be able to just do `&obj`, and that would be preferable.